File: DynamicPrograming/php/countPaths.php
<?php

declare(strict_types=1);

/** @param string[][] $grid */
function countPaths(array $grid): int
{
    $rows = count($grid);
    $cols = $rows ? count($grid[0]) : 0;
    $memo = [];

    $dfs = function (int $r, int $c) use (&$dfs, &$memo, $grid, $rows, $cols): int {
        if ($r >= $rows || $c >= $cols || $grid[$r][$c] === 'X') {
            return 0;
        }
        if ($r === $rows - 1 && $c === $cols - 1) {
            return 1;
        }

        $key = $r . ',' . $c;
        if (array_key_exists($key, $memo)) {
            return $memo[$key];
        }

        return $memo[$key] = $dfs($r + 1, $c) + $dfs($r, $c + 1);
    };

    return $dfs(0, 0);
}

File: DynamicPrograming/csharp/CountPaths.cs
using System.Collections.Generic;

namespace DynamicProgrammingSolutions;

public static class CountPaths
{
    public static int Count(string[][] grid)
    {
        var rows = grid.Length;
        var cols = rows == 0 ? 0 : grid[0].Length;
        var memo = new Dictionary<(int, int), int>();
        return Dfs(0, 0, grid, rows, cols, memo);
    }

    private static int Dfs(int r, int c, string[][] grid, int rows, int cols, IDictionary<(int, int), int> memo)
    {
        if (r >= rows || c >= cols || grid[r][c] == "X")
        {
            return 0;
        }

        if (r == rows - 1 && c == cols - 1)
        {
            return 1;
        }

        var key = (r, c);
        if (memo.TryGetValue(key, out var cached))
        {
            return cached;
        }

        var paths = Dfs(r + 1, c, grid, rows, cols, memo) + Dfs(r, c + 1, grid, rows, cols, memo);
        memo[key] = paths;
        return paths;
    }
}
